browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,139 @@
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
+
+  return domain;
+}
+
+// Session storage ids are expected to be 36 character long GUID strings. Since
+// we are constructing our own ids, we convert our string into a 32 character
+// hash and then use that make up our own GUID-like string. Because of the way
+// we are constructing the string we should never collide with a real GUID and
+// we only need to worry about hash collisions, which are unlikely.
+std::string StringToSessionStorageId(const std::string& string) {
+  std::string hash = base::MD5String(string) + "____";
+  DCHECK_EQ(hash.size(), 36u);
+  return hash;
+}
+
+}  // namespace
+
+PerTLDEphemeralStorage::PerTLDEphemeralStorage(
+    PerTLDEphemeralStorageKey key,
+    scoped_refptr<content::SessionStorageNamespace> local_storage_namespace)
+    : key_(key), local_storage_namespace_(local_storage_namespace) {
+  DCHECK(active_per_tld_storage_areas().find(key) ==
+         active_per_tld_storage_areas().end());
+  active_per_tld_storage_areas().emplace(key, weak_factory_.GetWeakPtr());
+}
+
+PerTLDEphemeralStorage::~PerTLDEphemeralStorage() {
+  active_per_tld_storage_areas().erase(key_);
+}
+
+EphemeralStorageTabHelper::EphemeralStorageTabHelper(WebContents* web_contents)
+    : WebContentsObserver(web_contents) {
+  DCHECK(base::FeatureList::IsEnabled(blink::features::kBraveEphemeralStorage));
+}
+
+EphemeralStorageTabHelper::~EphemeralStorageTabHelper() {}
+
+void EphemeralStorageTabHelper::ReadyToCommitNavigation(
+    NavigationHandle* navigation_handle) {
+  if (!navigation_handle->IsInMainFrame())
+    return;
+  if (navigation_handle->IsSameDocument())
+    return;
+
+  const GURL& new_url = navigation_handle->GetURL();
+  std::string new_domain = URLToStorageDomain(new_url);
+  std::string previous_domain =
+      URLToStorageDomain(web_contents()->GetLastCommittedURL());
+  if (new_domain == previous_domain)
+    return;
+
+  auto* browser_context = web_contents()->GetBrowserContext();
+  auto site_instance =
+      content::SiteInstance::CreateForURL(browser_context, new_url);
+  auto* partition =
+      BrowserContext::GetStoragePartition(browser_context, site_instance.get());
+
+  // If another EphemeralStorageTabHelper is already using a storage area for
+  // our TLD, we can use that.  Otherwise, we create a fresh storage area.
+  PerTLDEphemeralStorageKey key = std::make_pair(browser_context, new_domain);
+  auto it = active_per_tld_storage_areas().find(key);
+  if (it != active_per_tld_storage_areas().end()) {
+    DCHECK(!it->second.WasInvalidated());
+
+    // The map stores a weak pointer and we are upgrading it to a scoped_refptr
+    // here.
+    per_tld_ephemeral_storage_ = it->second.get();
+  } else {
+    std::string local_partition_id = new_domain + "/ephemeral-local-storage";
+    auto local_storage_namespace = content::CreateSessionStorageNamespace(
+        partition, StringToSessionStorageId(local_partition_id));
```

Comment:
I'm probably missing something, but if `CreateSessionStorageNamespace` returns same namespace for a given partition (which maps 1-1 to a browser context) and a given id, then we don't need a global map? Since storage partition already maintains one

Reply:
I believe that `CreateSessionStorageNamespace` returns a reference. If all the references disappear then, we the namespace is deleted. We need the map to maintain at least one reference, keeping the namespace alive.

Reply:
I'm honestly having hard times understanding this, will definitely need some help :) 
How does map helps us to keep a namespace alive if (1) it only holds weak pointers (2) last `per_tld_ephemeral_storage_` will erase the corresponding entry in the map?
So anyway, closing (or navigating away) all tabs with a given TLD cleans up the namespace

Reply:
The maps holds weak pointers, so the map alone won't keep the PerTLDEphemeralStorage (and the namespace) alive on its own. The TabHelper does keep a real reference to the PerTLDEphemeralStorage though. As long as one TabHelper has a reference, the PerTLDEphemeralStorage will stay alive. When the last TabHelper drops the reference (either by being destroyed or by switching domains during navigation), the PerTLDEphemeralStorage will be deleted by the reference counting code and the destructor removes it from the map of weak pointers.